### DISCUSSION TITLE
ORM/asyncio: Best way to get inserted ID?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#8576] [Usage Questions] [answered]
I'm just wondering what the preferred approach is for getting the inserted ID?

This is roughly what I was expecting to work:
```
thing = Thing(value="foo")
async with db.begin() as sess:
    sess.add(thing)
return thing.id
```

But, that gives `sqlalchemy.orm.exc.DetachedInstanceError: Instance <...> is not bound to a Session; attribute refresh operation cannot proceed`.

Seems I can do:
```
async with db.begin() as sess:
    sess.add(thing)
    await sess.flush()
    return thing.id
```
But, I thought the point of the `with` was to avoid needing the manual flushing.

I guess I was assuming that an insert statement would return an ID, and therefore be automatically populated on the DB object. So, the final line in my first example would be accessing the previously returned value from that object.

**Comment by zzzeek** [comment, marked as answer]
you would either do the second one, or you would have expire_on_commit=False on your Session and do the first one.  if you're in asyncio you probably want expire_on_commit=False.

in 2.0, you will also be able to reference INSERT more directly like this:

    thing = session.scalar(insert(Thing).returning(Thing), {"data": "some data"})

**Reply by Dreamsorcerer** [reply]
Perfect. I see there is also a dedicated `async_sessionmaker()` now. Maybe it would be worth defaulting to `False` in the function?

**Reply by zzzeek** [reply]
i think explicit is better than implicit on things like this, having it be the opposite would be worse IMO.  there is really not a good default for that parameter because it will do "the wrong thing" for someone no matter which way you flip it.
